firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationSquasher.java

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.model.mutation;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import com.google.firebase.Timestamp;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.FieldPath;
+import com.google.firebase.firestore.model.MutableDocument;
+import com.google.firebase.firestore.model.ObjectValue;
+import com.google.firebase.firestore.model.SnapshotVersion;
+import com.google.firestore.v1.Value;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Squashes multiple mutations applied on a document into a final mutation. It is guaranteed when
+ * the final mutation is applied on a document, it is as if the entire mutation list has been
+ * applied.
+ */
+public class MutationSquasher {
+
+  /** The key of the document whose mutations will be squashed */
+  private DocumentKey key;
+  /** Whether the document in question exists after `currentSquashed` is updated */
+  private boolean exists;
+  /** The mutation that is squashed from all the mutations it has seen. */
+  private Mutation currentSquashed = null;
+
+  public MutationSquasher(DocumentKey key, boolean exists) {
+    this.key = key;
+    this.exists = exists;
+  }
+
+  /** Returns the squashed mutation */
+  public Mutation getMutation() {
+    return currentSquashed;
+  }
+
+  /** Squashes the given mutation */
+  public void squash(Mutation mutation) {
+    if (mutation instanceof SetMutation) {
+      currentSquashed = mutation;
+      exists = true;
+    } else if (mutation instanceof DeleteMutation) {
+      currentSquashed = mutation;
+      exists = false;
+    } else if (currentSquashed == null) {
+      squashPatchMutationWithNull((PatchMutation) mutation);
```

Comment:
It is not obvious to me why this is a Patch (but I believe you). 

I am also a bit confused why this switches from checking the mutation type to checking `currentSquashed`.

BTW - what do you think of just calling these "overlays" instead of squashed mutations?